Cython/Compiler/ParseTreeTransforms.py

```diff
@@ -2202,12 +2201,79 @@ def visit_FuncDefNode(self, node):
     def visit_ClassDefNode(self, node):
         orig_qualified_name = self.qualified_name[:]
         entry = (getattr(node, 'entry', None) or             # PyClass
-                 self.current_env().lookup_here(node.name))  # CClass
+                 self.current_env().lookup_here(node.target.name))  # CClass
         self._append_entry(entry)
         self._super_visit_ClassDefNode(node)
         self.qualified_name = orig_qualified_name
         return node
 
+class MangleDunderNamesTransform(ScopeTrackingTransform):
+    """__names appearing inside classes are mangled to their _classname__names form
+
+    This should be called after the pure Python @cython directives have all been
+    applied"""
+    def visit_PyClassDefNode(self, node):
+        # The class name (target), list of bases and metaclass name should be
+        # processed without being mangled by the class itself.
+        # Note that for Python 2 the exclusion of the metaclass is incorrect
+        # since it is defined inside the class as __metaclass__
+        exclude_at_first = ['bases', 'metaclass', 'target']
```

Comment:
There is a mechanism for this that uses the list of `outer_attrs` for a node. Look for `visit_scoped_children`. Seems like the `ScopeTrackingTransform` should implement this as well. Having a special argument for a `visit_…()` method seems … surprising.

Reply:
I've solved this by no-longer inheriting from `ScopeTrackingTransform`. I wasn't actually using any features of it, except for the `last_class_name` attribute I added.